OneWayMultiplayerExample/Assets/Scripts/AI/TroopAI.cs

```diff
@@ -0,0 +1,414 @@
+        float dotProduct = Vector3.Dot(dirFromAtoB, transform.forward);
+        return dotProduct > 0.9f;
+    }
+    
+    //give direction to move troop towards
+    private void MoveTroop()
+    {
+        if (_target != null)
+        {
+            _navMeshAgent.destination = _target.transform.position;    
+        }
+        else
+        {
+            _navMeshAgent.destination = _homeLocation;
+        }
+    }
+    
+    //give a positive direction if you want to face the target and 
+    //a negative direction to look away
+    private void RotateToTarget()
+    {
+        //rotate enemy
+        Vector3 direction = Vector3.zero;
+        if (_target != null)
+        {
+            direction = (_target.transform.position - transform.position).normalized;
+        }
+        else
+        {
+            direction  = (_homeLocation - transform.position).normalized;
+        }
+        if (direction == Vector3.zero) return;
+        
+        _currQuat = Quaternion.LookRotation(direction);
+        transform.rotation = Quaternion.RotateTowards(transform.rotation, _currQuat, RotationSpeed * Time.deltaTime);
+    }
+    
+    //Try to make this a slottable action so this function never has to be overridden
+    private void PerformAction()
+    {
+        switch (EnemyType)
+        {
+            case EnemyTypes.Grunt:
+            case EnemyTypes.Soldier:
+                MeleeTarget();
+                break;
+            case EnemyTypes.Shooter:
+                StartShootingAnimation();
+                break;
+        }
+    }
+
+    private void MeleeTarget()
+    {
+        if (_isAttacking) return;
+        _isAttacking = true;
+        _animator.SetBool(IsAttacking, true);
+    }
+
+    private void StartShootingAnimation()
+    {
+        _animator.SetBool(IsAttacking, true);
+    }
+    
+    //This function is triggered through an Animation Event in Unity
+    public void ShootTarget()
+    {
+        if (_shootScript && gameObject && _target)
+        {
+            _shootScript.SpawnProjectile(gameObject.layer, _target);
+        }
+    }
+
+    /// <param name="in_teamID">0 = invader(local player), 1 = defender(network player)</param>
+    public void AssignToTeam(int in_teamID)
+    {
+        GameManager.Instance.Troops.Add(this);
+        TeamID = in_teamID;
+        if (in_teamID == 0)
+        {
+            //Invaders
+            _activeMask = DefenderMask;
+            if (_meleeWeapon)
+            {
+                _meleeWeapon.gameObject.layer = 6; 
+            }
+            //6 = Invader Layer, 7 = Defender Layer
+            gameObject.layer = 6;
+            _healthBar.AssignTeamColor(Color.blue);
+        }
+        else
+        {
+            //Defenders
+            _activeMask = InvaderMask;
+            if (_meleeWeapon)
+            {
+                _meleeWeapon.gameObject.layer = 7; 
+            }
+            //6 = Invader Layer, 7 = Defender Layer
+            gameObject.layer = 7;
```

Comment:
It would be nice to have the layers be a public const in a static class if they're going to be set in scripts.